### DISCUSSION TITLE
PrepareAndDrawmolecule function doesn't work for me

### DISCUSSION THREAD

**Discussion by TAEUKHAM** [rdkit/rdkit#4743] [Q&A] [answered]
Hello, I'm trying to highlight typical molecule with designated spot and found "PrepareAndDrawmolecule" function could do this for me.
However, this function seems not working for me even if I've copy and pasted the sample code from https://www.rdkit.org/docs/GettingStartedInPython.html
where I've tried to find out the function work.

This is the tutorial code.
Thanks
```
from rdkit import Chem
from rdkit.Chem.Draw import rdMolDraw2D

smi = 'c1cc(F)ccc1Cl'
mol = Chem.MolFromSmiles(smi)
patt = Chem.MolFromSmarts('ClccccF')
hit_ats = list(mol.GetSubstructMatch(patt))
hit_bonds = []
for bond in patt.GetBonds():
    aid1 = hit_ats[bond.GetBeginAtomIdx()]
    aid2 = hit_ats[bond.GetEndAtomIdx()]
    hit_bonds.append(mol.GetBondBetweenAtoms(aid1,aid2).GetIdx())

colours = [(0.8,0.0,0.8),(0.8,0.8,0),(0,0.8,0.8),(0,0,0.8)]
atom_cols = {}

for i, at in enumerate(hit_ats):
    atom_cols[at] = colours[i%4]
bond_cols = {}
for i, bd in enumerate(hit_bonds):
    bond_cols[bd] = colours[3 - i%4]

d = rdMolDraw2D.MolDraw2DCairo(500, 500)
rdMolDraw2D.PrepareAndDrawMolecule(d, mol, highlightAtoms=hit_ats,
                                   highlightAtomColors=atom_cols,
                                   highlightBonds=hit_bonds,
                                   highlightBondColors=bond_cols)
```

**Comment by greglandrum** [comment, marked as answer]
Hi @TAEUKHAM : what do you mean by "it doesn't work for me"? Please include the error/warning messages you see and/or what you think the code is doing which is incorrect.

**Reply by TAEUKHAM** [reply]
@greglandrum Hi, thanks for the reply. In the rdkit document, it says "PrepareAndDrawMolecule function preps a molecule for drawing and actually draws it". However when I call this function, it returns nothing with no error. 

It should draw this picture when the sample code runs
![20211129_150737](https://user-images.githubusercontent.com/88373558/143817228-e4f6076a-1ec6-4ca3-aeff-6c174b6f4408.png)
.

**Reply by greglandrum** [reply]
I think you've misunderstood what that code does (the documentation should be clearer about this). `PrepareAndDrawMolecule()` does end up drawing the molecule using the Draw2D object you pass it, but it doesn't return the image itself to you.
If you want to see the image in the notebook then you need to do something like:
```
from IPython.display import Image
d.FinishDrawing()
png = d.GetDrawingText()
Image(png)
```
Here's a gist showing the whole thing:
https://gist.github.com/greglandrum/5f6e541a376e96c2049344db6371b76b

**Reply by TAEUKHAM** [reply]
wow, thanks. I've expected PrepareAndDrawMolecule() to draw image immediately like lots of drawing function in Chemdraw library does:) Now it works fine:)
